globus_cli/parsing/endpoint_plus_path.py

```diff
@@ -0,0 +1,63 @@
+import click
+
+
+class EndpointPlusPath(click.ParamType):
+    """
+    Custom type for "<endpoint_id>:<path>"
+    Supports path being required and path being optional.
+
+    Always produces a Tuple.
+    """
+    name = 'endpoint plus path'
+
+    def __init__(self, *args, **kwargs):
+        # path requirement defaults to True, but can be tweaked by a kwarg
+        self.path_required = kwargs.pop('path_required', True)
+
+        super(EndpointPlusPath, self).__init__(*args, **kwargs)
+
+    def get_metavar(self, param):
+        """
+        Default metavar for this instance of the type.
+        """
+        return self.metavar
+
+    @property
+    def metavar(self):
+        """
+        Metavar as a property, to satisfy the slight brokenness of
+        click.Argument
+        Tracked against Click as an issue:
+        https://github.com/pallets/click/issues/674
+        """
+        if self.path_required:
+            return "ENDPOINT_ID:PATH"
+        else:
+            return "ENDPOINT_ID[:PATH]"
+
+    def convert(self, value, param, ctx):
+        """
+        ParamType.convert() is the actual processing method that takes a
+        provided parameter and parses it.
+        """
+        # passthrough conditions: None or already processed
+        if value is None or isinstance(value, tuple):
+            return value
+
+        # split the value on the first colon, leave the rest intact
+        splitval = value.split(':', 1)
+        # first element is the endpoint_id
+        endpoint_id = click.UUID(splitval[0])
```

Comment:
Returning `endpoint_id` as a UUID class seems to breaks things, e.g. with `async-delete`, when it tries to activate the endpoint.

Python 2:
```
  File "build/bdist.linux-x86_64/egg/globus_cli/parsing/main_command_decorator.py", line 25, in invoke
  File "/home/mayor/Repos/globus-cli/.venv2/lib/python2.7/site-packages/click-6.6-py2.7.egg/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/mayor/Repos/globus-cli/.venv2/lib/python2.7/site-packages/click-6.6-py2.7.egg/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/mayor/Repos/globus-cli/.venv2/lib/python2.7/site-packages/click-6.6-py2.7.egg/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/mayor/Repos/globus-cli/.venv2/lib/python2.7/site-packages/click-6.6-py2.7.egg/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "build/bdist.linux-x86_64/egg/globus_cli/services/transfer/async_delete.py", line 41, in async_delete_command
  File "build/bdist.linux-x86_64/egg/globus_cli/services/transfer/activation.py", line 6, in autoactivate
  File "/home/mayor/Repos/globus-cli/.venv2/lib/python2.7/site-packages/globus_sdk-0.4.2-py2.7.egg/globus_sdk/transfer/client.py", line 286, in endpoint_autoactivate
    path = self.qjoin_path("endpoint", endpoint_id, "autoactivate")
  File "/home/mayor/Repos/globus-cli/.venv2/lib/python2.7/site-packages/globus_sdk-0.4.2-py2.7.egg/globus_sdk/base.py", line 95, in qjoin_path
    return "/" + "/".join(quote(part) for part in parts)
  File "/home/mayor/Repos/globus-cli/.venv2/lib/python2.7/site-packages/globus_sdk-0.4.2-py2.7.egg/globus_sdk/base.py", line 95, in <genexpr>
    return "/" + "/".join(quote(part) for part in parts)
  File "/usr/lib64/python2.7/urllib.py", line 1297, in quote
    if not s.rstrip(safe):
AttributeError: 'UUID' object has no attribute 'rstrip'
```

Python 3:
```
  File "/home/mayor/Repos/globus-cli/.venv3/lib/python3.5/site-packages/globus_cli-0.4.2.0-py3.5.egg/globus_cli/parsing/main_command_decorator.py", line 25, in invoke
  File "/home/mayor/Repos/globus-cli/.venv3/lib/python3.5/site-packages/click-6.6-py3.5.egg/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/mayor/Repos/globus-cli/.venv3/lib/python3.5/site-packages/click-6.6-py3.5.egg/click/core.py", line 1060, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/mayor/Repos/globus-cli/.venv3/lib/python3.5/site-packages/click-6.6-py3.5.egg/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/mayor/Repos/globus-cli/.venv3/lib/python3.5/site-packages/click-6.6-py3.5.egg/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/home/mayor/Repos/globus-cli/.venv3/lib/python3.5/site-packages/globus_cli-0.4.2.0-py3.5.egg/globus_cli/services/transfer/async_delete.py", line 41, in async_delete_command
  File "/home/mayor/Repos/globus-cli/.venv3/lib/python3.5/site-packages/globus_cli-0.4.2.0-py3.5.egg/globus_cli/services/transfer/activation.py", line 6, in autoactivate
  File "/home/mayor/Repos/globus-cli/.venv3/lib/python3.5/site-packages/globus_sdk-0.4.2-py3.5.egg/globus_sdk/transfer/client.py", line 286, in endpoint_autoactivate
    path = self.qjoin_path("endpoint", endpoint_id, "autoactivate")
  File "/home/mayor/Repos/globus-cli/.venv3/lib/python3.5/site-packages/globus_sdk-0.4.2-py3.5.egg/globus_sdk/base.py", line 95, in qjoin_path
    return "/" + "/".join(quote(part) for part in parts)
  File "/home/mayor/Repos/globus-cli/.venv3/lib/python3.5/site-packages/globus_sdk-0.4.2-py3.5.egg/globus_sdk/base.py", line 95, in <genexpr>
    return "/" + "/".join(quote(part) for part in parts)
  File "/usr/lib64/python3.5/urllib/parse.py", line 712, in quote
    return quote_from_bytes(string, safe)
  File "/usr/lib64/python3.5/urllib/parse.py", line 737, in quote_from_bytes
    raise TypeError("quote_from_bytes() expected bytes")
TypeError: quote_from_bytes() expected bytes
```